### PR TITLE
Fix OpenLoco#388: Re-center Options window on Scale Factor change

### PR DESCRIPTION
**Pull request by seifer7**
Easy fix by closing and re-opening the Options window after the Scale Factor change has been applied.

I preferred this method over trying to re-position the window to the center by working out the X/Y positions.

Why? Because this is using the already existing `open` function to position the window exactly where it's expected (the center of the screen in this case, but if that ever got changed then this method won't need changing as it's using the existing `open` function)

### CODE REVIEW

**Comment by janisozaur** [comment]
The code looks ok, but I'd like for this to be tested as well.

Are you interested in addressing https://github.com/OpenLoco/OpenLoco/issues/418 too?

**Comment by seifer7** [comment]
> 
> The code looks ok, but I'd like for this to be tested as well.

I tested on the title screen and in-game and seems to work as expected. Obviously someone else should double check my work.

> 
> Are you interested in addressing #418 too?

Unfortunately this is beyond my abilities (at least for now.) I've never worked in C++ til now.

**Comment by AaronVanGeffen** [comment]
Hmm. I'm not sure I like the idea of closing and reopening just for the net effect of centring the window. I'd prefer windows to have a `moveToCentre()` member function. This function would have to do three things:

1. Invalidate the screen for the old window position (`window->invalidate()`)
2. Reposition the window to the centre
3. Invalidate the screen for the new window position

**Comment by seifer7** [comment]
@AaronVanGeffen How about this?

Implement new function `WindowManager::moveToCentre`

New function can be called in three different ways, and returns a `window*` or `nullptr` if window doesn't exist.
If multiple windows of the same type exist, the first is moved and it is returned.

Calling methods:
moveToCentre(WindowType type);
moveToCentre(WindowType type, window_number number);
moveToCentre(window* w);

File: src/openloco/ui/WindowManager.cpp [hunk 1]
```diff
@@ -623,6 +623,44 @@ namespace openloco::ui::WindowManager
...
+        }
+        return nullptr;
+    }
+
+    window* moveToCentre(window* w)
+    {
+        w->invalidate();
+        w->x = ((ui::width() - w->width) / 2);
+        w->y = ((ui::height() - w->height) / 2);
+        gfx::invalidate_screen();
```
Comment (AaronVanGeffen) [inline src/openloco/ui/WindowManager.cpp:660]: You really shouldn't need to invalidate the entire screen. Any particular reason you're not calling `invalidate` again?

**Comment by AaronVanGeffen** [comment]
Conceptually, this is fine. I would have preferred a member function on the window itself (i.e. on the window struct), though. Personally, I don't think we'll need the first two signatures, but they probably won't hurt either.

@marijnvdwerf Do you have a preference w.r.t. having this in WindowManager or Window?

**Comment by seifer7** [comment]
I hadn't looked at the window struct, my bad.

I will move it into the Window struct, and also use the existing Window::move function which takes care of the invalidation.

**Comment by AaronVanGeffen** [comment]
Sounds like a plan. Thanks!

**Comment by seifer7** [comment]
Okay done. I've added a member function `moveToCentre` to the `window` struct.

This is now called after the "Window Scale" option is changed to move the Options window back to the centre of the screen.

This new function can now be used any time a Window needs to be moved to the centre of the screen.
`window->moveToCentre();`

File: src/openloco/window.cpp [hunk 2]
```diff
@@ -604,6 +604,16 @@ namespace openloco::ui
         }
     }
 
+    bool window::moveToCentre()
+    {
+        float dx = ((ui::width() - this->width) / 2);
+        float dy = ((ui::height() - this->height) / 2);
```
Comment (AaronVanGeffen) [inline src/openloco/window.cpp:610]: No need to use floats here; in fact, decimal precision is not desirable for pixels here. Note that window::move also takes two int16_t as its parameters.
```suggestion
        int16_t dx = ((ui::width() - this->width) / 2);
        int16_t dy = ((ui::height() - this->height) / 2);
```

**Comment by AaronVanGeffen** [comment]
This looks much more compact, very nice. I've left a minor comment that I'd like to see addressed, though.

**Pull request merged**